Modules/_queuemodule.c

```diff
@@ -25,12 +26,158 @@ static struct PyModuleDef queuemodule;
 
+static const Py_ssize_t kInitialRingBufCapacity = 8;
+
+typedef struct {
+    // Where to place the next item
+    Py_ssize_t put_idx;
+
+    // Where to get the next item
+    Py_ssize_t get_idx;
+
+    PyObject **items;
+
+    // Total number of items that may be stored
+    Py_ssize_t items_cap;
+
+    // Number of items stored
+    Py_ssize_t num_items;
+} RingBuf;
+
+static int
+RingBuf_Init(RingBuf *buf)
+{
+    buf->put_idx = 0;
+    buf->get_idx = 0;
+    buf->items_cap = kInitialRingBufCapacity;
+    buf->items = PyMem_Calloc(buf->items_cap, sizeof(PyObject*));
+    buf->num_items = 0;
+    return buf->items == NULL ? -1 : 0;
+}
+
+static PyObject *
+RingBuf_At(RingBuf *buf, Py_ssize_t idx)
+{
+    assert(idx >= 0 && idx < buf->num_items);
+    return buf->items[(buf->get_idx + idx) % buf->items_cap];
+}
+
+static void
+RingBuf_Fini(RingBuf *buf)
+{
+    PyObject **items = buf->items;
+    Py_ssize_t num_items = buf->num_items;
+    Py_ssize_t cap = buf->items_cap;
+    Py_ssize_t idx = buf->get_idx;
+    buf->items = NULL;
+    buf->put_idx = 0;
+    buf->get_idx = 0;
+    buf->num_items = 0;
+    buf->items_cap = 0;
+    for (Py_ssize_t n = num_items; n > 0; idx = (idx + 1) % cap, n--) {
+        Py_CLEAR(items[idx]);
+    }
+    PyMem_Free(items);
+}
+
+static void
+coalesce_items(RingBuf *buf, PyObject **new_items, Py_ssize_t new_capacity)
+{
+    // Copy the "tail" of the old items array. This corresponds to "head" of
+    // the abstract ring buffer.
+    Py_ssize_t tail_size = Py_MIN(buf->num_items, buf->items_cap - buf->get_idx);
+    if (tail_size > 0) {
+        memcpy(new_items, buf->items + buf->get_idx, tail_size * sizeof(PyObject *));
+    }
+
+    // Copy the "head" of the old items array, if any. This corresponds to the
+    // "tail" of the abstract ring buffer.
+    Py_ssize_t head_size = buf->num_items - tail_size;
+    if (head_size > 0) {
+        memcpy(new_items + tail_size, buf->items, head_size * sizeof(PyObject *));
+    }
+
+    PyMem_Free(buf->items);
+    buf->items = new_items;
+    buf->items_cap = new_capacity;
+    buf->get_idx = 0;
+    buf->put_idx = buf->num_items;
+}
+
+static void
+shrink_ringbuf(RingBuf *buf)
+{
+    Py_ssize_t new_capacity = Py_MAX(kInitialRingBufCapacity, buf->items_cap / 2);
+    assert(new_capacity >= buf->num_items);
+    if (new_capacity == buf->items_cap) {
+        return;
+    }
+
+    PyObject **new_items = PyMem_Calloc(new_capacity, sizeof(PyObject *));
+    if (new_items == NULL) {
+        // It's safe to ignore the failure; shrinking is an optimization and
+        // isn't required for correctness.
+        return;
+    }
+
+    coalesce_items(buf, new_items, new_capacity);
+}
+
+// Returns an owned reference
+static PyObject *
+RingBuf_Get(RingBuf *buf)
+{
+    assert(buf->num_items > 0);
+
+    if (buf->num_items < (buf->items_cap / 2)) {
+        // Items is less than 50% occupied, shrink it
+        shrink_ringbuf(buf);
+    }
```

Comment:
Yeah I was concerned about this too and was chatting with @DinoV about whether or not we should use a lower threshold here. I opted not to because I believe this matches the behavior of the previous implementation. Happy to switch to 25%.

Reply:
This matches the list shrinking behavior, but list growth is slower. For large sizes it grows by 1.125x instead of 2x, so it doesn't risk the quadratic behavior.